// lit-frontend/app/swing-video-container.ts
import { css, html, LitElement } from "lit";
import { customElement, property } from "lit/decorators.js";

@customElement("swing-video-container")
export class SwingVideoContainerElement extends LitElement {
    @property({ reflect: true, type: Boolean })
    open: boolean = false;

    render() {
        return html`
        <div class="swing-video-container">
            <div class="filter">
                <label for="filter">Sort By</label>
                <select name="filter">
                    <option value="date-asc">Date: Most Recent</option>
                    <option value="date-desc">Date: Least Recent</option>
                </select>
            </div>
            <slot></slot>
        </div>
    `;
    }

    static styles = css`
    .swing-video-container{
        display: flex;
        flex-direction: row;
        align-items: flex-start;
        flex-wrap: wrap;
    }

    .filter {
      display: flex;
      font-size: 20px;
      background-color: white;
      box-sizing: border-box;
      width: 100%;
      border-radius: 8px;
      align-items: center;
      padding: 10px;
      margin: 0 74px 0 40px;
    }

    label{
        padding: 5px;
    }

    select{
        font-size: 16px;
    }

  `;
}
